pkg/resource/restconfig_factory.go

```diff
@@ -41,16 +148,27 @@ func (r *restConfigClientGetter) ToDiscoveryClient() (discovery.CachedDiscoveryI
 
 // ToRESTMapper returns a restmapper
 func (r *restConfigClientGetter) ToRESTMapper() (meta.RESTMapper, error) {
-	discoveryClient, err := r.ToDiscoveryClient()
-	if err != nil {
-		return nil, err
+	// HIVE-2399: store the mapper and its discovery client for future use
+	r.mu.Lock()
+	if r.mapper == nil {
+		r.logger.Debugf("creating discovery client and mapper for ToRESTMapper()")
+		discoveryClient, err := r.ToDiscoveryClient()
+		if err != nil {
+			r.mu.Unlock()
+			return nil, err
+		}
+		r.mapper = restmapper.NewDeferredDiscoveryRESTMapper(discoveryClient)
+		r.discoveryClient = discoveryClient
+	} else {
+		r.logger.Debugf("reusing discovery client and mapper for ToRESTMapper()")
 	}
+	r.mu.Unlock()
```

Comment:
Use `defer r.mu.Unlock()` after L152 instead. If you're really wanting to keep the mutex window tight, wrap L152-164 in an anonymous `func(){}()`.